### PR TITLE
Windows node graceful shutdown

### PR DESCRIPTION
**Pull request by zylxjtu**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This is to implement the windows node shutdown, which is proposed in here:
https://github.com/kubernetes/enhancements/pull/4813

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Enabled graceful shutdown feature for Windows node
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by hashim21223445 — APPROVED** [review]

**Review by hashim21223445 — APPROVED** [review]

**Comment by kannon92** [comment]
Please sign the CLA.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_others.go
Comment (marosset) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_others.go:2]: nit: update the comment below

> // NewManager returns a fake node shutdown manager for non linux platforms.

Maybe just say "for unsupported platforms"?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (marosset) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:89]: Let's add a log statement here

File: pkg/windows/service/service.go
Comment (marosset) [inline pkg/windows/service/service.go:99]: Does this need a comment too?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_others.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_others.go:2]: updated

File: pkg/windows/service/service.go
Comment (zylxjtu) [reply pkg/windows/service/service.go:99]: will add

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:89]: Will add

**Review by marosset — APPROVED** [review]
/lgtm
Thanks!

File: pkg/windows/service/service.go
Comment (aravindhp) [inline pkg/windows/service/service.go:66]: Are you sure that other error cases don't need to be handled?

**Review by aravindhp — COMMENTED** [review]
Thanks for working on this @zylxjtu. Looks good. I just have a couple of questions.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:214]: the changes we made to update the registry is for the kublet dependency shutdown order. I think there is no need to revert it back.

I'm wondering maybe we can these few lines above the 206?

File: pkg/windows/service/service.go
Comment (zylxjtu) [reply pkg/windows/service/service.go:66]: it should be good enough, an example code is here: https://go.googlesource.com/sys/+/master/windows/svc/mgr/config.go#71

**Review by aravindhp — COMMENTED** [review]
/lgtm

**Comment by dims** [comment]
/milestone v1.32

We have a feature gate here from sig-windows that needs to land in v1.32

File: test/e2e/windows/node_shutdown.go
Comment (SergeyKanzhelev) [inline test/e2e/windows/node_shutdown.go:47]: Why we need those `MinimumKubeletVersion` and `WindowsDocker`? Neither is supported upstream for a very long time. So we can safely assume this test is not running on either

File: test/e2e/windows/node_shutdown.go
Comment (SergeyKanzhelev) [inline test/e2e/windows/node_shutdown.go:355]: the issue is fixed. Why code needs to reference it?

File: test/e2e/windows/node_shutdown.go
Comment (SergeyKanzhelev) [inline test/e2e/windows/node_shutdown.go:134]: Is this signal recoverable? Can you reset it somehow on Windows after it was set? Or effectively you can only run a single test per execution?

If it can be reset - I suggest to have `AfterEach` that will reset the signal and restart kubelet to clean up state after each test

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:57]: I forgot about this file completely. Filing a bug to mention it on: https://kubernetes.io/docs/reference/node/kubelet-files/?

Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:57]: https://github.com/kubernetes/website/issues/48610

Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:57]: is there some logic duplication around this file? We use the same on Linux, right?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:338]: is this function reentrant?  I don't quite understand what it does, but wonder if there are any issues with it if kubelet was restarted multiple times.

Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:338]: from the callsite it may look like the list will keep growing every kubelet reastart. But I might be wrong

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows_test.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows_test.go:287]: please add re-entry tests. Kubelet restart should not be growing this list and ideally should just keep it as-is.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:211]: do we have to set it if it's the same as before?

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:213]: also add old value here

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:219]: why we log from here, but not from other error returns below in this same function?

**Review by SergeyKanzhelev — COMMENTED** [review]
non-windows code changes are OK. There is a sign off from SIG Windows, I left a few comments there, but it is not like I understand the system aspects of it very well.

File: test/e2e/windows/node_shutdown.go
Comment (zylxjtu) [reply test/e2e/windows/node_shutdown.go:47]: This is copied from an old template, will remove

File: test/e2e/windows/node_shutdown.go
Comment (zylxjtu) [reply test/e2e/windows/node_shutdown.go:355]: This is still in the linux related test: https://github.com/kubernetes/kubernetes/blob/88a997ac01cd8a386df9318d81b89257307f53fb/test/e2e_node/node_shutdown_linux_test.go#L254. I can remove it if this is no longer an issue

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:57]: Thanks! Please let me know if will need me to do anything.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows_test.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows_test.go:287]: will add one

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:211]: yes, this is the current logic, otherwise we will need to create new slices and do the comparison, I'm thinking maybe a blind copy is easier and performant

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:219]: Will update, I guess the extra logging mostly the leftover from my initial debugging

**Comment by drewhagen** [comment]
👋 @zylxjtu @SergeyKanzhelev @aravindhp @mrunalp @saschagrunert @marosset @derekwaynecarr Thanks for contributing and reviewing this PR.

⚠️ I see this PR is labeled ❗️priority/important-soon and tagged for the `v1.32` release.

Therefore, I want to extend a friendly reminder that the code freeze is starting `02:00 UTC Friday November 8th 2024` (a little less than **1 week from now**). Please help make sure the PR has both `lgtm` and `approved` labels before the code freeze. Thanks! 🚀

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:215]: will the logging the old value, in the lines above

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:339]: It will not add reduandent values. This has also been verified with e2e tests, besides the unit test cases

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:57]: if you want to help with that doc, it will be helpful.

File: pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:220]: ping when updated.

Comment (zylxjtu) [reply pkg/kubelet/nodeshutdown/nodeshutdown_manager_windows.go:220]: @SergeyKanzhelev , already send out a new iteration with force update, please help to take a look

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

Most of my approve is relying on sig windows's approval.

File: pkg/windows/service/service.go
Comment (SergeyKanzhelev) [inline pkg/windows/service/service.go:21]: I would suggest to put an OWNERS file in this directlory so SIG Windows can approve it.

If you do so, please also include SIG Node team as approvers

Comment (zylxjtu) [reply pkg/windows/service/service.go:21]: thanks, this is also what we discussed before, let me check with @marosset and @jsturtevant

Comment (SergeyKanzhelev) [reply pkg/windows/service/service.go:21]: Please make sure to add Node reviewers and approvers into that file since it will be Node-related code. I don't think we should block every change in this file by `pkg/` approvers.

Comment (zylxjtu) [reply pkg/windows/service/service.go:21]: Added Owner file @SergeyKanzhelev

File: pkg/windows/service/OWNERS
Comment (SergeyKanzhelev) [inline pkg/windows/service/OWNERS:5]: use alias `sig-node-approvers` instead of explicit names

File: pkg/windows/service/OWNERS
Comment (SergeyKanzhelev) [inline pkg/windows/service/OWNERS:4]: why reviewers have approvals permission? Is it intentional?

File: pkg/windows/service/OWNERS
Comment (SergeyKanzhelev) [inline pkg/windows/service/OWNERS:11]: also add labels:
```suggestion
  - sig-node-reviewers
  
 
labels:
  - sig/windows
  - sig/node
```

Comment (zylxjtu) [reply pkg/windows/service/OWNERS:11]: added

File: pkg/windows/service/OWNERS
Comment (SergeyKanzhelev) [inline pkg/windows/service/OWNERS:8]:
```suggestion
  - sig-node-approvers
```

Comment (SergeyKanzhelev) [reply pkg/windows/service/OWNERS:8]: @marosset anybody from wig windows explicitly?

File: pkg/windows/service/OWNERS
Comment (SergeyKanzhelev) [reply pkg/windows/service/OWNERS:11]: thx. I suggest just do sig-node-approvers as suggested above and merge it already. We can add Windows approvers later

Comment (zylxjtu) [reply pkg/windows/service/OWNERS:8]: updated

Comment (SergeyKanzhelev) [reply pkg/windows/service/OWNERS:8]: pls remove `sig-windows-api-reviewers` from approvers

File: pkg/windows/service/OWNERS
Comment (marosset) [reply pkg/windows/service/OWNERS:4]: Let's move all the edits to the OWNERS file out to a seperate PR

Comment (SergeyKanzhelev) [reply pkg/windows/service/OWNERS:4]: I didn't expect it will be such a struggle. I am OK with the separate PR

File: pkg/windows/service/OWNERS
Comment (zylxjtu) [reply pkg/windows/service/OWNERS:7]: updated

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by SergeyKanzhelev** [comment]
/assign @dims 

@dims are you ok to approve the new OWNERS file?

**Pull request merged**
